File: spotifyarch/src/main/java/com/timothy/spotifyarchitecture/livedata/TokenLiveData.java
package com.timothy.spotifyarchitecture.livedata;

import android.arch.lifecycle.LiveData;
import android.content.Context;
import android.support.annotation.MainThread;

import com.timothy.spotifyarchitecture.entities.Token;
import com.timothy.spotifyarchitecture.remote.Resource;

/**
 * Created by tim on 6/4/17.
 */

public class TokenLiveData extends LiveData<Resource<Token>> {
    private static TokenLiveData tInstance;

    @MainThread
    public static TokenLiveData get() {
        if (tInstance == null)
            tInstance = new TokenLiveData();
        return tInstance;
    }

}

File: spotifyarch/src/main/java/com/timothy/spotifyarchitecture/livedata/TokenLiveData.kt
package com.timothy.spotifyarchitecture.livedata

import android.arch.lifecycle.MediatorLiveData
import com.timothy.spotifyarchitecture.notNullEmpty
import com.timothy.spotifyarchitecture.retrofit.models.Token

/**
 * LiveData implementation to create a MediatorLiveData<Token>
 */

class TokenLiveData private constructor() : MediatorLiveData<Token>() {

    companion object {
        @JvmField
        var tInstance: TokenLiveData = TokenLiveData()

        @JvmStatic
        fun needsRefresh(): Boolean {
	        return notNullEmpty(tInstance.value?.refresh_token)
			        && System.currentTimeMillis() >= tInstance.value !!.expires_in
        }
    }

}
